File: Assets/Scripts/Core/Effects/SwapTraitEffect.cs
using System.Linq;

namespace Bunker.Core.Effects
{
    // Swaps one still-hidden trait category between the caster and the target.
    // Both players must have that category unrevealed, so the swap stays
    // invisible to everyone else until either of them chooses to reveal it.
    public class SwapTraitEffect : ISpecialCardEffect
    {
        public bool CanApply(GameSession session, PlayerData caster, PlayerData target)
        {
            if (target == null || target.IsEliminated) return false;
            if (target.PlayerId == caster.PlayerId) return false;

            // At least one category must be hidden for both players.
            return caster.Traits.Any(t =>
                !caster.IsCategoryRevealed(t.Category) &&
                !target.IsCategoryRevealed(t.Category));
        }

        // Category to swap is chosen by the caster; passed in via GameSession
        // as an extra parameter (see GameSession.UseSpecialCard overload below),
        // since ISpecialCardEffect.Apply's signature is intentionally generic.
        public void Apply(GameSession session, PlayerData caster, PlayerData target)
        {
            // Concrete swap logic lives in GameSession.UseSpecialCard because it
            // needs the extra "which category" parameter. This method exists
            // to satisfy the interface and is intentionally left as a no-op here.
        }
    }
}

File: bunker-server/src/game/effects/SwapTraitEffect.ts
import { ISpecialCardEffect } from "./ISpecialCardEffect";
import { GameSession } from "../GameSession";
import { PlayerData } from "../PlayerData";

// Swaps one still-hidden trait category between the caster and the target.
// Category selection happens in GameSession (extra parameter needed) —
// this canApply only checks that at least one swappable category exists.
export class SwapTraitEffect implements ISpecialCardEffect {
  canApply(_session: GameSession, caster: PlayerData, target: PlayerData | null): boolean {
    if (!target || target.isEliminated) return false;
    if (target.playerId === caster.playerId) return false;

    return caster
      .getTraits()
      .some((t) => !caster.isCategoryRevealed(t.category) && !target.isCategoryRevealed(t.category));
  }

  apply(_session: GameSession, _caster: PlayerData, _target: PlayerData | null): void {
    // No-op here — actual swap logic lives in GameSession.useSwapTraitSpecialCard,
    // which has the extra "which category" parameter this interface doesn't carry.
  }
}
